### ISSUE TITLE
Cannot convert Mnesia to PostgreSQL with erl2sql:export

### ISSUE THREAD

**Issue by nekoswag** [processone/ejabberd#2232]
> What version of ejabberd are you using?

17.12

> What operating system (version) are you using?

openSUSE Leap 42.3

> How did you install ejabberd (source, package, distribution)?

distribution

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

I'm trying to migrate from the mnesia database backend to PostgreSQL.
I created a PostgreSQL user and database "ejabberd", and imported this schema:
https://github.com/processone/ejabberd/blob/master/sql/pg.new.sql
Added the PostgreSQL config as described here:
https://docs.ejabberd.im/admin/configuration/#database-and-ldap-configuration

Started an Erlang shell with "ejabberdctl debug", and entered this:
`ejd2sql:export(<<"mydomain.com">>, sql).`

It says "ok", but error.log shows this:

`2018-01-17 23:28:12.194 [error] <0.383.0>@ejabberd_sql:outer_transaction:504 SQL transaction restarts exceeded
** Restarts: 10
** Last abort reason: [{severity,'FEHLER'},{86,<<"ERROR">>},{code,<<"23502">>},{message,<<"NULL-Wert in Spalte Â»server_hostÂ« verletzt Not-Null-Constraint">>},{detail,<<"Fehlgeschlagene Zeile enthÃ¤lt (user3, null, 04Ul0eb7Zr/Pm3DuwztJeB5cWtI=, iGei20ZycnTgRedTJMlyxaLMxzo=, v5/RIV8CJdxav8EVKl3sJA==, 4096, 2018-01-17 23:28:12.193487).">>},{115,<<"public">>},{116,<<"users">>},{99,<<"server_host">>},{file,<<"execMain.c">>},{line,1731},{routine,<<"ExecConstraints">>}]
** Stacktrace: [{ejabberd_sql,sql_query_t,1,[{file,"src/ejabberd_sql.erl"},{line,186}]},{lists,foreach,2,[{file,"lists.erl"},{line,1337}]},{ejabberd_sql,outer_transaction,3,[{file,"src/ejabberd_sql.erl"},{line,497}]},{ejabberd_sql,run_sql_cmd,4,[{file,"src/ejabberd_sql.erl"},{line,434}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]
** When State == {state,<0.404.0>,pgsql,90606,30000,<<"mydomain.com">>,{{[],[]},0,unlimited}}
2018-01-17 23:28:12.195 [error] <0.534.0>@ejd2sql:export:95 Failed export for module ejabberd_auth and table passwd: {{badmatch,{aborted,[{severity,'FEHLER'},{86,<<"ERROR">>},{code,<<"23502">>},{message,<<"NULL-Wert in Spalte Â»server_hostÂ« verletzt Not-Null-Constraint">>},{detail,<<"Fehlgeschlagene Zeile enthÃ¤lt (user3, null, 04Uler37Zr/Pm3DuPkwwcy5cWtI=, iGei20ZpviTgRedTJMlyxaLMxzo=, v5/RIV8CJddeyxEVKlwsJA==, 4096, 2018-01-17 23:28:12.193487).">>},{115,<<"public">>},{116,<<"users">>},{99,<<"server_host">>},{file,<<"execMain.c">>},{line,1731},{routine,<<"ExecConstraints">>}]}},[{ejd2sql,output,4,[{file,"src/ejd2sql.erl"},{line,186}]},{mnesia_tm,apply_fun,3,[{file,"mnesia_tm.erl"},{line,833}]},{mnesia_tm,execute_transaction,5,[{file,"mnesia_tm.erl"},{line,808}]},{ejd2sql,'-export/3-fun-4-',5,[{file,"src/ejd2sql.erl"},{line,87}]},{lists,foreach,2,[{file,"lists.erl"},{line,1337}]},{ejd2sql,export,3,[{file,"src/ejd2sql.erl"},{line,85}]},{lists,foreach,2,[{file,"lists.erl"},{line,1337}]},{ejd2sql,export,2,[{file,"src/ejd2sql.erl"},{line,71}]}]}
2018-01-17 23:28:12.470 [error] <0.385.0>@ejabberd_sql:outer_transaction:504 SQL transaction restarts exceeded
`
[snip]
[usernames and domain replaced by me]

"NULL-Wert in Spalte "server_host" verletzt Not-Null-Constraint"
basically means
"NULL-value in column "server_host" violates not-null-constraint"

and a whole bunch of other errors like one that says
"insertion or update in table "pubsub_state" violates foreign key-constraint "pubsub_state_nodeid_fkey""

Is it perhaps related to https://github.com/processone/ejabberd/issues/2183 ?
So I'm starting to think this binary has no PostgreSQL support... How can I check if it does?

**Comment by cromain** [comment]
You should create schema with pg.sql instead of pg.new.sql, as your distribution likely did not compiled ejabberd with option --enable-new-sql-schema.

**Comment by JonnyDevid** [comment]
where should I place this command ( --enable-new-sql-schema. ) in windows cmd?

**Comment by licaon-kter** [comment]
@JonnyDevid It's probably already included if you use the provided .EXE installer.

Do you have the same issue?

**Comment by JonnyDevid** [comment]
yes i do, i want migrating mnesia to pgsql o mysql but no found, I can not find a step-by-step tutorial by migrating.
